<a href="https://colab.research.google.com/github/testowankopythonika2/Machine-Learning-Python-Bootcamp/blob/main/Regresja_liniowa_r%C3%B3wnanie_normalne.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Aby zainstalować bibliotekę scikit-learn, użyj polecenia poniżej:

In [1]:
!pip install scikit-learn

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


### Aby zaktualizować do najnowszej wersji bibliotekę scikit-learn, użyj polecenia poniżej:

In [ ]:
!pip install --upgrade scikit-learn

### Import bibliotek

In [2]:
import numpy as np

### Przygotowanie danych

In [3]:
X1 = np.array([1, 2, 3, 4, 5, 6])
Y = np.array([3000, 3250, 3500, 3750, 4000, 4250])
m = len(X1)

print(f'Lata pracy: {X1}')
print(f'Wynagrodzenie: {Y}')
print(f'Liczba próbek: {m}')

Lata pracy: [1 2 3 4 5 6]
Wynagrodzenie: [3000 3250 3500 3750 4000 4250]
Liczba próbek: 6


In [4]:
X1 = X1.reshape(m, 1)
print(X1)
print(X1.shape)

[[1]
 [2]
 [3]
 [4]
 [5]
 [6]]
(6, 1)


In [6]:
bias = np.ones((m, 1))
print(bias)
print(bias.shape)

[[1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]]
(6, 1)


In [7]:
X = np.append(bias, X1, axis=1)
print(X)
print(X.shape)

[[1. 1.]
 [1. 2.]
 [1. 3.]
 [1. 4.]
 [1. 5.]
 [1. 6.]]
(6, 2)


### Równanie normalne

Regresja liniowa w R2:

                                    Y = w0 + w1 * X1
                                    Y = W^T*X
gdzie:
                                    W = [w0\nw1]
                                    X = [1\nX1]

stąd
                                  W^T = [w0, w1]
                          Y = W^T*X = [w0, w1] * [1\nX1] = w0 + w1 * X1

Równanie normalne - równanie pozwalające obliczyć minimum funkcji straty (o ile istnieje)
                                  W  = (X^T * X)^-1(X^T*Y)

In [8]:
np.dot(X.T, X)

array([[ 6., 21.],
       [21., 91.]])

In [9]:
L = np.linalg.inv(np.dot(X.T, X))
L

array([[ 0.86666667, -0.2       ],
       [-0.2       ,  0.05714286]])

In [10]:
P = np.dot(X.T, Y)
P

array([21750., 80500.])

In [11]:
np.dot(L, P)

array([2750.,  250.])

Końcowa postać modelu:

                        Y = 2750 + 250X1

### Regresja liniowa przy pomocy scikit-learn

In [12]:
from sklearn.linear_model import LinearRegression

regression = LinearRegression()
regression.fit(X1, Y)

print(regression.intercept_)
print(regression.coef_[0])

2750.0
250.0
